src/testing/burgers_energy_conservation_rrk.cpp

```diff
@@ -0,0 +1,161 @@
+int BurgersEnergyConservationRRK<dim, nstate>::compare_energy_to_initial(
+        const std::shared_ptr <DGBase<dim, double>> dg,
+        const double initial_energy,
+        bool expect_conservation
+        ) const{
+    
+    const double final_energy = compute_energy_collocated(dg);
+    const double energy_change = abs(initial_energy-final_energy);
+    pcout << "Energy change at end was " << energy_change <<std::endl;
+    if (expect_conservation && (energy_change< 1E-13)){
+        pcout << "Energy was conserved, as expected." << std::endl;
+        return 0; //pass test
+    } else if (!expect_conservation && (energy_change > 1E-13)){
+        pcout << "Energy was NOT conserved, as expected." << std::endl;
+        return 0; //pass test
+    }else if (expect_conservation && (energy_change > 1E-13)){
+        pcout << "Energy was NOT conserved, but was expected to be conserved." << std::endl;
+        pcout << "    Unexpected result! Test failing." << std::endl;
+        return 1; //fail test
+    }else{
+        pcout << "Energy was conserved, but was expected NOT to be conserved." << std::endl;
+        pcout << "    Unexpected result! Test failing." << std::endl;
+        return 1; //fail test (included for completeness, but not expected to be used)
+    }
+}
+
+template <int dim, int nstate>
+double BurgersEnergyConservationRRK<dim, nstate>::compute_energy_collocated(
+        const std::shared_ptr <DGBase<dim, double>> dg
+        ) const
+{
+    double energy = 0.0;
+    for (unsigned int i = 0; i < dg->solution.size(); ++i)
+    {
+        energy += 1./(dg->global_inverse_mass_matrix.diag_element(i)) * dg->solution(i) * dg->solution(i);
+    }
+    return energy;
+}
+
+template <int dim, int nstate>
+int BurgersEnergyConservationRRK<dim,nstate>::run_flow_solver(
+        const Parameters::AllParameters params,
+        const double energy_initial,
+        bool expect_conservation
+        ) const
+{
+    std::unique_ptr<FlowSolver::FlowSolver<dim,nstate>> flow_solver = FlowSolver::FlowSolverFactory<dim,nstate>::select_flow_case(&params, parameter_handler);
+    static_cast<void>(flow_solver->run());
+    int failed_this_calculation = compare_energy_to_initial(flow_solver->dg, energy_initial, expect_conservation);
+    return failed_this_calculation;
+}
+
+template <int dim, int nstate>
+int BurgersEnergyConservationRRK<dim, nstate>::run_test() const
+{
+
+    double final_time = this->all_parameters->flow_solver_param.final_time;
+    double time_step_large = this->all_parameters->ode_solver_param.initial_time_step;
+    double time_step_small = time_step_large * 1E-2;
+
+    int n_steps = round(final_time/time_step_large);
+    if (n_steps * time_step_large != final_time){
+        pcout << "Error: final_time is not evenly divisible by initial_time_step!" << std::endl
+              << "Remainder is " << fmod(final_time, time_step_large)
+              << ". Modify parameters to run this test." << std::endl;
```

Comment:
Instead of having to modify the parameters manually, could you create a parameters object thats a copy of  `this->all_parameters` called `adjusted_parameters`, adjust the final_time so that it does satisfy this criteria, and pass `adjusted_parameters` when creating the flow solver

Reply:
and add a pcout saying that the time step was adjusted

Reply:
I think it's safer to keep it as-is, to leave it up to the user's discretion whether their chosen time-step size or their chosen final time is more important. If I made the change you're suggesting, some scenarios would have unintended behaviour - for example, if the user is trying to stop calculation just before a shock forms and has made a mistake in entering the timestep.